---
title: "Label ROC Data"
author: "Michael A. Gilchrist"
date: "24 Jul 2020"
output: pdf_document
---

# Preliminary Information
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

To compile in emacs use M-n e 


# Purpose
Use information from FASTA file used to fit ROC to add isoform.ID and WormBase.ID information

# Load Libraries
```{r}
library(Biostrings) ## process first to avoid conflicts
library(tidyr)
library(tibble)
library(readr)
library(dplyr)
library(stringr)
library(forcats)
library(ggplot2)
library(knitr)
library(ggpubr)
library(ggpmisc)
library(purrr)

creationInfo <- paste0("\tDate: ", date(), "\n\tLocation: ", sub(".*/AcrossTissue", "AcrossTissue", getwd()))

exportData=TRUE ## Flag for running save(), write.csv() and other output commands
```

# Load and Shape Data

## ROC 

### Load FASTA to get isoform and WormBase IDs (names)

```{r, eval = exportData}
## Load WB FASTA file
## Using Biostrings function which is not a standard df
seqData <- readDNAStringSet("Input/c_elegans.PRJNA13758.WS277.CDS_transcripts.fa")

seqLength <- width(seqData)

## names are really long descriptions.
## NEed to extract relevant part
seqDesc  <- as_tibble_col(x=names(seqData), column_name = "desc")
seqID <- separate(data = seqDesc, col = desc, into=c("isoform.ID", "WormBase.ID"), sep=" gene=", remove=TRUE) 
rm(seqDesc)

## sub example
##WormBase.ID <- sub(".* gene=([WBGEne0-9]+)", "\\1", seqDesc)



## verify there's a match for each entry
if(sum(is.na(seqID$WormBase.ID)) ==0) print("Every entry matches")

## Verify that all 'names' are unique.
if(length(seqID$WormBase.ID) != length(unique(seqID$WormBase.ID))) print("Some WormBase IDs appear twice due to isoforms")

if(length(seqID$isoform.ID) == length(unique(seqID$isoform.ID))) print("Every isoform.ID is unique as expected.")

```

## Label ROC estimates
Load ROC Estimates and bind isoform.ID and WormBase.ID with Phi Values
```{r}
## Import Phi Values from ROC Output

## detailed information on phi: posterior mean, posterior mean of log10(phi), etc
## StdError really StdDev of posterior
## This will change with a ROC update
unlabeledROCOutput <-
    readr::read_csv("Input/ROC_unlabeled.phi.summaries.with.sphi.equal.2.8.csv") 

labeledROCOutput <- bind_cols(seqID, unlabeledROCOutput) 

write.csv(x=labeledROCOutput, file="Output/ROC_labeled.phi.summaries.with.sphi.equal.2.8.csv", quote=FALSE, row.names=FALSE) 
```

## No longer needed...


```{r}
dplyr::rename(phi = PHI, log10Phi = log10.PHI, sd = Std.Dev, sdLog10Phi = log10.Std.Dev) %>%
    mutate(logPhi=log10Phi*log(10), sdLogPhi=sdLog10Phi*log(10)) %>%
    select(c(phi, sd, log10Phi, sdLog10Phi))

phiMatchData  <- as_tibble(bind_cols(WormBase.ID, phiPosteriorInfo, length=seqLength))


dim(phiMatchData)

## Problem: Isoforms of the same gene exist in the fasta file (and thus phi estimates), but are not part of the E-MTAB measurements which are indexed the gene, not isoform of a gene, level.
## Previous Solution: Combine separate estimates using mean or median values of phi - See 21_Process.Published.Means/load.and.process.means.data.Rmd
## Current Solution: 
summaryStatsPhiData  <- phiData %>% group_by(geneName) %>% summarize(mean_phi = mean(phi), mean_sd = mean(sd), mean_var = mean(sd^2), mean_length = mean(length), mean_logPhi=mean(logPhi), mean_sdLogPhi=mean(sdLogPhi), mean_logVar=mean(sdLogPhi^2), median_phi = median(phi), median_sd = median(sd), median_length = median(length), median_logPhi=median(logPhi), median_sdLogPhi=median(sdLogPhi), n_isoforms = length(phi) )

comment(summaryStatsPhiData) <- "summary stats for means and sd of phi for a geneName's multiple isoforms which are summarized in n_isoforms"
dim(summaryStatsPhiData)

write.csv(unfilteredPhiData, file="Output/labeled.unfiltered.phi.data-NCBI.csv", quote=FALSE, row.names=FALSE)
write.csv(unfilteredFastaInfoNCBI, file="Output/unfiltered.fasta.info-NCBI.csv", quote= FALSE, row.names=FALSE)
write.csv(anomolousGeneInfo, "Output/anomolousGeneInfo-NCBI.csv", quote=FALSE, row.names=FALSE)
write.csv(summaryStatsPhiData, file="Output/summaryStatsPhiData-NCBI.csv", quote=FALSE, row.names=FALSE)
save(phiData, seqData, summaryStatsPhiData, file = "Output/processed.ROC.data-NCBI.Rdata")

```

# Look at distribution of isoforms and filter
```{r}
nameCounts <- count(as_tibble(WormBase.ID), value)
qplot(x=n, data=nameCounts)

geneNamesSingleIsoforms <- filter(nameCounts, n ==1) 
## Filter out whose length is not a multiple of 3
## Keep name of filtered data simple
unfilteredFastaInfoNCBI <- tibble(info=names(unfilteredSeqData), geneName=unfilteredSeqGeneNames, length=as.double(unfilteredSeqLength))

## Create a vector flagging isoforms with proper gene lengths
properLengthIsoformsFlag <- ((unfilteredFastaInfoNCBI$length %% 3) == 0)

## Check for MtDNA genes
## Results indicate there are none. Good!

## note seqData uses 'names' while fastaInfoNCBI uses 'info'
seqData <- unfilteredSeqData[properLengthIsoformsFlag, ]
fastaInfoNCBI <- unfilteredFastaInfoNCBI[properLengthIsoformsFlag, ]


```

Import data if `exportData==FALSE`
```{r eval = !(exportData)}
## Load data instead of generating and exporting it
unfilteredPhiData <- read.csv(file="Output/labeled.unfiltered.phi.data-NCBI.csv")
unfilteredFastaInfoNCBI <-  read.csv(file="Output/unfiltered.fasta.info-NCBI.csv")
anomolousGeneInfo <- read.csv("Output/anomolousGeneInfo-NCBI.csv")
summaryStatsPhiData <- read.csv(file="Output/summaryStatsPhiData-NCBI.csv")
load(file = "Output/processed.ROC.data-NCBI.Rdata")
```


### Plot Results
```{r}
## Examine Unfiltered and Filtered Data
dim(unfilteredPhiData)
plot(unfilteredPhiData$phi, unfilteredPhiData$sd,
     main="Plot Includes 'anomolous' isoforms",
     sub = "anomolous = gene lengths in nts that are not multiples of 3"
     )

## Verify we've filtered correctly
plot(phiData$phi, phiData$sd, main="Plot excludes 'anomolous' isoforms",
     sub = "anomolous = gene lengths in nts that are not multiples of 3"
     )

plot(anomolousPhiData$phi, anomolousPhiData$sd, main="Isoforms Whose Lengths are Not Multiples of Three")

#par(mfrow=c(2,2))
plot(summaryStatsPhiData$mean_phi, summaryStatsPhiData$mean_sd)
mtext("Summary Statistics of Phi Estimates", side = 3, line = -21, outer = TRUE)
plot(summaryStatsPhiData$mean_length, summaryStatsPhiData$mean_sd, log="xy")
plot(summaryStatsPhiData$mean_phi, summaryStatsPhiData$mean_length, log="xy")
## Create histogram of short genes
tmp <- summaryStatsPhiData$mean_length[ summaryStatsPhiData$mean_length < 400]
hist(tmp, xlim=c(0, max(tmp)), main="Histogram of Gene Lengths < 400 aa", xlab="Length" )



## Log metrics
#par(mfrow=c(2,2))
plot(summaryStatsPhiData$mean_logPhi, summaryStatsPhiData$mean_sdLogPhi)
mtext("Summary Statistics of log(Phi) Estimates", side = 3, line = -21, outer = TRUE)
plot(summaryStatsPhiData$mean_length, summaryStatsPhiData$mean_sdLogPhi, log="x")
plot(summaryStatsPhiData$mean_logPhi, summaryStatsPhiData$mean_length, log="y")

```

