---
title: "Fit Error in Variables Model to Stage Specific RNAseq Data and ROC Phi"
author: "Michael A. Gilchrist"
date: "21 Jul 2020"
output: pdf_document
---

# ESS Commands
See (~/Software/R/r.notes.Rmd)  for more details

| Keys    | Description              |
|:--------|:-------------------------|
| M-n v v | evaluate current chunk   |
| C-M-x   | evaluate function/region |
| M-n e   | evaluate entire buffer   |


# Purpose
- Fit stage specific measurements of gene expression to phi values estimated by ROC.

# Results So Far
- Have restricted genes in the 20% to 80%-ile of the phi values estimated by ROC.
  Based on Alex Cope suggested restricting data to genes where the counts are 'not low', e.g. > 100 in every stage if we were working with FPKM (as I understand it).
- Initially using simple polynomial models.
	- Initial fits were poor
		- negative coefficients
		- highly structured residuals that declined linearly with the predictor.
		(See figure for linear `lm()` fit)
	- Residuals suggest a second order polynomial will work well.
- Could see if log(phi) is better predicted using the geometric mean (though a 0 would mess things up).
- May eventually fit error in variables regression approach since there is substantial error in both predictor (RNAseq counts) and response variable (ROC's $\phi$).





# Load Libraries
```{r}
##library(Biostrings) ## process first to avoid conflicts
## May want to set library(verbose=TRUE)
library(tidyr)
library(tibble)
library(readr)
library(dplyr)
library(stringr)
library(forcats)
library(ggplot2)
library(knitr)
library(ggpubr)
library(ggpmisc)
library(gridExtra)
library(purrr)
library(eivtools)

creationInfo <- paste0("\tDate: ", date(), "\n\tLocation: ", sub(".*/AcrossTissue", "AcrossTissue", getwd()))

#exportData=TRUE ## Flag for running save(), write.csv() and other output commands
```



# Load and Filter Data


```{r}
## Define file names
emtabDataFile <- "Input/processed.E-MTAB.data.Rdata"
isoformSummaryStatsFile <- "Input/ROC.isoform.summary.stats-NCBI.Rdata"
embryoStageCountMomentsFile <- "Input/summary.stats.of.embryo.stage.data.csv"


##labeledPhiFile <- "Input/ROC_labeled.phi.summaries.with.sphi.equal.2.8.csv"

## Load E-MTAB Data
load(file=emtabDataFile) ## lifeStageCount and lifeStage

lifeStageCount <- rename(lifeStageCount,
                         WormBase.ID = WBID)

## $\phi$ Data
load(isoformSummaryStatsFile)
comment(isoformSummaryStats)


```

## Plot ROC Isoform Data
- Should probably reside some where else
- May already exist in Plot.ROC...Rmd files
```{r eval=TRUE}
## remove unnecessary data on phi
## In future may want to use WBID if it is available
## Since there is more than one isoform/gene, don't use phiData.
## Instead use summary stats of phi data



selectStats  <- select(isoformSummaryStats, c("WormBase.ID", "mean_phi", "mean_var"))


## only retain genes with matches in both datasets
## NOTE:
meanLog10MeanPhi <- mean(log10(selectStats$mean_phi))
sdLog10MeanPhi <- sd(log10(selectStats$mean_phi))

qplot(data = selectStats, x=mean_phi, log = "x") # +
#    annotate(geom = geom_vline(aes(xintercept = meanLog10MeanPhi) ) ) #+ 
#    annotate(geom = geom_segment(x = meanLog10MeanPhi, xend = meanLog10MeanPhi + sdLog10MeanPhi, y = 0, yend=0) )

qplot(data = selectStats, x=mean_var, log = "y")

```






# Reshape E-MTAB Data
```{r}

## filteredWBID  <- hLifeStageCount %>% group_by(WBID) %>% summarise( nStages = sum(count > threshold)) %>% filter(nStages > 0) %>% select(WBID)

## Remove count and totalCount because it messes with the results of pivot_wider()
tallData <- lifeStageCount %>%
    filter(grepl("hermaphrodite", sex)) ##%>%
    ## Ensure average for each set of data is 1
    ## Should update scaledCount later when genes are filtered based on Phi values
##    mutate(scaledCount=(count/sum(count)*), totalCount=sum(count)) %>%
##    group_by(stage) %>%
##    semi_join(filteredWBID) %>% select(-c(tissue, sex, count, totalCount))

wideData <- pivot_wider(tallData, names_from=stage, values_from=count )

## Update column names
names(wideData) <-
    names(wideData) %>%
    str_replace_all(' Ce', '') %>%
    str_replace_all(' larva', '') %>%
    str_replace_all(' stage', '') %>%
    str_replace_all(' ', '.)'

```

```{r}

## Find genes in common between ROC and E-MTAB data
commonGenesData <- inner_join(wideData, wtPhiData, by = "geneName")


commonGenesData
```


# Fit model


## Filter Genes with Extreme Phi Values
```{r}
myPhiPercentileRange <- c(0.2, 0.8))
filterPhiRange <- quantile(x = selectStats[["mean_phi"]], probs= myPhiPercentileRange)

filteredGeneData <- filter(commonGenesData, mean_phi > filterPhiRange[1] & mean_phi < filterPhiRange[2])

```

```{r} 



qplot(x = mean_phi, y = weight_phi, data = commonGenesData)

lmUnweightedPhiFit  <- lm(mean_phi ~ embryo+L1+L2+L3+L4+adult+dauer+post.dauer-1, filteredGeneData)


plot(lmUnweightedPhiFit)
summary(lmUnweightedPhiFit)
## Need to do a contrained fit where
## sum(coefficients) =1 and coefficients > 0



```
