---
title: "Load and Process Dat from E-MTAB files and ROC"
author: "Michael A. Gilchrist"
date: "12 Jul 2020"
output: html_document
---

# Preliminary Information
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. 

# Purpose
The purpose of this document is to process empirical measurements of mRNA abundances using mRNA-seq technologies.
Data is in 
These measurements are based on multiple measurements.
The values are generally the means of the counts.
Unfortunately we don't know anything about the sd of the values used to calculate the mean counts.


# Load Libraries
```{r}
library(Biostrings) ## process first to avoid conflicts
library(tidyr)
library(tibble)
library(readr)
library(dplyr)
library(stringr)
library(forcats)
library(ggplot2)
##library(hash)
##library(purrr)
creationInfo <- paste0("\tDate: ", date(), "\n\tLocation: ", sub(".*/AcrossTissue", "AcrossTissue", getwd()))

```

# Load and Shape Data

## ROC 

### Get gene names
#### Extract from FASTA file
- There are a bunch of anomolous genes that have a phi of 1 but an elevated sd relative to the rest of genes with that value
Turns out they have a width which is not a multiple of 3.
- Note that a gene may have some isoforms that are not multiples of 3, so we don't want to exclude all isoforms of a gene.
- Data that hasn't been filtered for these anomolous genes has the prefix 'unfiltered'.
- Filtered data does not have a prefix.
- ROC analyses should be rerun with these genes filtered out.

```{r}
## Load original FASTA file
## Using Biostrings function which is not a standard df
unfilteredSeqData <- readDNAStringSet("Input/c_elegan.fasta")

## names are really long descriptions.
## NEed to extract relevant part
unfilteredSeqDesc  <- names(unfilteredSeqData)
unfilteredSeqLength <- width(unfilteredSeqData)
unfilteredSeqGeneNames <- sub(".*\\[gene=([^]]+).*", "\\1", unfilteredSeqDesc)
##seqGeneWidth<- width(seqData) %>% as_tibble()

## use geneName for descript
#names(unfilteredSeqData) <- unfilteredSeqGeneNames

## verify there's a match for each entry
if(sum(is.na(unfilteredSeqGeneNames)) ==0) print("Every entry matches")

## Verify that all 'names' are unique.
if(length(unfilteredSeqGeneNames) != length(unique(unfilteredSeqGeneNames))) print("Some geneNames appear twice due to isoforms")


## Filter out whose length is not a multiple of 3
## Keep name of filtered data simple
unfilteredFastaInfo <- tibble(info=names(unfilteredSeqData), geneName=unfilteredSeqGeneNames, length=as.double(unfilteredSeqLength))

## Create a vector flagging isoforms with proper gene lengths
properLengthIsoformsFlag <- ((unfilteredFastaInfo$length %% 3) == 0)

## Check for MtDNA genes
## Results indicate there are none. Good!
## filter(unfilteredFastaInfo, grepl("MtDNA", info) )

## note seqData uses 'names' while fastaInfo uses 'info'
seqData <- unfilteredSeqData[properLengthIsoformsFlag, ]
fastaInfo <- unfilteredFastaInfo[properLengthIsoformsFlag, ]
```




#### Import Phi Values from ROC Output
```{r}
## detailed information on phi: posterior mean, posterior mean of log10(phi), etc
## StdError really StdDev of posterior
unfilteredPhiPosteriorInfo <- readr::read_csv("Input/phi.posterior.unlabeled.csv") %>% dplyr::rename(phi = PHI, logPhi = log10.PHI, sd = Std.Error, logSd = log10.Std.Error)

## Not yet filtered for genes that have anomolous lengths (i.e. length mod 3 !=0)
unfilteredPhiData  <- as_tibble(bind_cols(geneName=unfilteredSeqGeneNames, unfilteredPhiPosteriorInfo, length=unfilteredSeqLength))
write.csv(unfilteredPhiData, file="Output/labeled.unfiltered.phi.data.csv", quote=FALSE, row.names=FALSE)

# rm(unfilteredPhiPosteriorInfo)
## 
dim(unfilteredPhiData)
plot(unfilteredPhiData$phi, unfilteredPhiData$sd,
     main="Plot Includes 'anomolous' isoforms",
     sub = "anomolous = gene lengths in nts that are not multiples of 3"
     )
phiData <- unfilteredPhiData[properLengthIsoformsFlag, ]

## verify you've filtered correctly
plot(phiData$phi, phiData$sd, main="Plot excludes 'anomolous' isoforms",
     sub = "anomolous = gene lengths in nts that are not multiples of 3"
     )

anomolousLengthIsoformsFlag <- !properLengthIsoformsFlag
anomolousGeneInfo <- unfilteredFastaInfo[anomolousLengthIsoformsFlag, ]
anomolousPhiData  <-unfilteredPhiData[anomolousLengthIsoformsFlag, ]
plot(anomolousPhiData$phi, anomolousPhiData$sd, main="Isoforms Whose Lengths are Not Multiples of Three")
write.csv(anomolousGeneInfo, "Output/anomolousGeneInfo.csv", quote=FALSE, row.names=FALSE)


## Problem: Isoforms of the same gene exist in the fasta file (and thus phi estimates), but are not part of the E-MTAB measurements which are indexed the gene, not isoform of a gene, level.
## Solution: Combine separate estimates using mean or median values of phi

summaryStatsPhiData  <- phiData %>% group_by(geneName) %>% summarize(mean_phi = mean(phi), mean_sd = mean(sd), mean_length = mean(length), mean_logPhi=mean(logPhi), mean_logSd=mean(logSd), median_phi = median(phi), median_sd = median(sd), median_length = median(length), median_logPhi=median(logPhi), median_logSd=median(logSd), n_isoforms = length(phi) )

comment(summaryStatsPhiData) <- "summary stats for means and sd of phi for a geneName's multiple isoforms which are summarized in n_isoforms"
dim(summaryStatsPhiData)

write.csv(summaryStatsPhiData, file="Output/summaryStatsPhiData.csv", quote=FALSE, row.names=FALSE)

save(phiData, summaryStatsPhiData, seqData, file = "Output/processed.ROC.data.Rdata") 


dev.off()

par(mfrow=c(2,2))
plot(summaryStatsPhiData$mean_phi, summaryStatsPhiData$mean_sd)
mtext("Summary Statistics of Phi Estimates", side = 3, line = -21, outer = TRUE)
plot(summaryStatsPhiData$mean_length, summaryStatsPhiData$mean_sd, log="xy")
plot(summaryStatsPhiData$mean_phi, summaryStatsPhiData$mean_length, log="xy")
## Create histogram of short genes
tmp <- summaryStatsPhiData$mean_length[ summaryStatsPhiData$mean_length < 400]
hist(tmp, xlim=c(0, max(tmp)), main="Histogram of Gene Lengths < 400 aa", xlab="Length" )



## log metrics
par(mfrow=c(2,2))
plot(summaryStatsPhiData$mean_logPhi, summaryStatsPhiData$mean_logSd)
mtext("Summary Statistics of log(Phi) Estimates", side = 3, line = -21, outer = TRUE)
plot(summaryStatsPhiData$mean_length, summaryStatsPhiData$mean_logSd, log="x")
plot(summaryStatsPhiData$mean_logPhi, summaryStatsPhiData$mean_length, log="y")
## Create histogram of short genes
tmp <- summaryStatsPhiData$mean_length[ summaryStatsPhiData$mean_length < 400]
hist(tmp, xlim=c(0, max(tmp)), main="Histogram of Gene Lengths < 400 aa", xlab="Length" )

```


#### Import Phi Values Using Lu's File -- BROKEN! NOT USED
- Don't use Lu's file
```{r eval=FALSE}

## NOTE: rocNames is corrupt.  It has replaced some gene names with date formats (e.g. apr-1 has been converted to 1-Apr
rocNamesBroken <- read_csv("Input/lu.phi.mean.by.names.csv", col_names = c("geneName","phi2") )
rocNamesBroken[rocNamesBroken$geneName=='1-Apr',]
## THere are 17 corrupted names
length(unfilteredSeqGeneNames) - sum(rocNamesBroken==unfilteredSeqGeneNames)

## Need to run 'Get estimates from ROC' code below to evaluate following line
phiBrokenDataCheck  <- bind_cols(rocNamesBroken, unfilteredPhiPosteriorInfo)
## Verify that phi values line up between the two datasets
plot(phiBrokenDataCheck$phi2, phiDataBrokenCheck$phi)
```


### Examine Anomolous Genes
- Originally observed as cluster of genes with phi~1 and elevated SDs
- Upon examination we see they are actually genes with n mod 3 !=0.
- Don't need to evaluate this code any more as of 12 Jul 2020
```{r eval=FALSE}
length(anomolousGeneNames$geneName)
length(unique(anomolousGeneNames$geneName))
anomolousInfo <- filter(unfilteredFastaInfo, geneName %in% anomolousGeneNames$geneName)
regularInfo <- filter(unfilteredFastaInfo, !(geneName %in% anomolousGeneNames$geneName))
```


## E-MTAB
### Load Data
```{r}
emtabFile <- "Input/E-MTAB-2812-query-results.tpms.tsv"
flatData  <- readr::read_tsv(emtabFile, skip=4) %>%
    dplyr::rename(WBID = `Gene ID`, geneName=`Gene Name` ) ## WBID is the WormBaseID

## use pivot_longer command (not gather which is depricated)
tmpData <- flatData %>% pivot_longer(-c(WBID, geneName), names_to = "long_description", values_to = "count")


## Now separate entry in descriptor column into separate column entries and filter out tissue specific data and genes that are not in the fasta file
tmpFastaGeneName <- unique(fastaInfo$geneName)
countData  <- separate(tmpData, long_description, into=c("sex", "tissue", "stage"), sep=", ") %>% filter(tissue=="organism") %>%
    filter(geneName %in% tmpFastaGeneName) %>%
    replace_na(list(count = 0)) # replace NAs with 0.
## Before doing so the min value in the tibble was 0.1

comment(countData) <- paste0("Data from ", emtabFile, "in tidy format. Empty cells filled with '0', genes not appearing in c_elegans.fasta have been removed out\n", creationInfo) 
```

### Create stageCounts 
- Data includes a new set of measurements for sex='hermaphrodite' stage='embryo Ce'.
This data is created from the mean values of all sex = 'hermaphrodite' and stage = '* embryo Ce' data
```{r}
## Get hermphrodite embryo stage names
## NB: male, 'embryo Ce' lacks a space at the start
embryoStageName  <- unique(countData$stage[grepl(' embryo Ce', countData$stage )])

separateEmbryoStageCount  <-  filter(countData, stage %in% embryoStageName) 

## 
## Make things simple and focus on just the 'main' lifestage measurements
## note that the stage "embryo Ce" is specific to sex = male
## Thus need to combine embryo measurements some how
##
## Only include data columns well find elsewhere
## See separate chunk for looking at moments of data
embryoStageCount <- separateEmbryoStageCount %>%
 group_by(WBID) %>%
    summarise(geneName, sex, tissue, stage="embryo Ce", count =mean(count), .groups="drop") %>%
    unique()


## embryoStageCount <- select(tmpEmbryoStageCount, -c(count, sd, tissue) ) %>% group_by()

nonEmbryoStageCount <- filter(countData, !(stage %in% embryoStageName)) %>% select(-tissue)

stageCount <- bind_rows(embryoStageCount, nonEmbryoStageCount)
comment(stageCount) <-  paste0("Data from ", emtabFile, "in tidy format. Empty cells filled with '0'\n", "separate  stages for embryonic hermaphrodites have been combined using mean(counts) for a WBID across embryonic stages\n", creationInfo) 
```

### Plot Hermaphrodite Embryo Data
```{r}
 ## Be sure to use count = func(count) last in summarise().
 ## Otherwise it will redefine count to a single value which will screw up the other functions
 ## Use mean() rather than median for summarizing count
 ## It seems less sensitive to 0 values
embryoStageCountMoments <- separateEmbryoStageCount %>%
    group_by(WBID) %>%
    summarise(mean=mean(count), inverseMean = 1/mean(count), median=median(count), sd=sd(count), n = length(count), min(count), max(count), .groups="drop") %>% unique()

embryoStageCountMoments %>%
    #filter(mean > 1) %>%
    ggplot() + geom_point(aes(mean, sd^2)) +
    scale_x_log10() +
    scale_y_log10() +
    labs(title="Summary Stats of Hermaphrodite Embryo Stages",
         x="mean", y="variance")

print("Data suggests log(var) = c log(mean) across biological replicates")
```

### Create Counts for mainLifeStages 
```{r}
## This will include both sex = c('male', 'hermaphrodite')\
## Really only lacks 'newly molted young adult hermaphrodite Ce' 
lifeStages <- c("embryo Ce", "L1 larvae Ce", "L2 larvae Ce", "L3 larvae Ce", "L4 larvae Ce", "adult Ce", "dauer larvae Ce", "post dauer stage Ce")
comment(lifeStages)=paste0("List of main lifestages to use in regression.\n", creationInfo)

## Note that 'count' are standardized means of biological and technical replicates
## 'organism' scale data for different lifestages
## Includes data for 'sex' =  male and hermphrodite
lifeStageCount <- filter(stageCount, (stage %in% lifeStages)) 
comment(lifeStageCount) <- c("RNASeq counts for each of the\n\t- hermaphrodite non-embryo lifestages: L1 larvae Ce, L2 larvae Ce, ... adult\n\t- male lifestages\n\t- mean of the very specific 'hermphrodite,  *embryo Ce' which are 'hermphrodite, embryo Ce.\nAll data for tissue == organism.\nCounts from E-MTAB-2812-query-results.tpms.tsv.", creationInfo ) 


lifeStageCountMoments <- lifeStageCount %>% group_by(stage)%>% summarize(stage_sd = sd(count), stage_logSd = sd(log(count, 10)) )


summary(lifeStageCount$count)
## hist(lifeStageCount$count)
hist(log(lifeStageCount$count))

geneNameAndWormBaseID <- unique(select(lifeStageCount, c(geneName, WBID)))
comment(geneNameAndWormBaseID) <- paste("Mapping between WormBase IDs and geneName in FASTA file")

#tidyGeneNameLifeStageStdCounts <- select(lifeStageCount, -WBID)
#tidyWBIDLifeStageStdCounts <- select(lifeStageCount, -geneName)

## Export values
## Don't use write_csv because that only works with dataframes,not tibbles
## write.csv converts the tibble to a dataframe
```

## Create database linking geneName and WormBaseID
Evaluate these lines manually outside of knitr
```{r eval=FALSE}
save(lifeStageCount, lifeStages, file = "Output/processed.E-MTAB.data.Rdata")
## Export data using 'geneName' name or WormBaseID
write.csv(geneNameAndWormBaseID, file="Output/geneName.and.WBID.csv", quote=FALSE, row.names=FALSE)
```


## Combine empirical E-MTAB and ROC data
```{r}
## Combine phi
#tidyFilteredAndWeightedLifeStageStdCount <- tidyLifeStageStdCount %>%
#    filter(geneName %in% phiData$geneName) %>%
#    mutate(weight=(1/(count+1)) ) ## x+1 poisson weighting


phiNames <- summaryStatsPhiData$geneName

namesToWBID <- lapply(phiNames, function(x) {index = which(geneNameAndWormBaseID$geneName == x); ifelse(x==0, NA, geneNameAndWormBaseID$WBID[index])})#, USE.NAMES=FALSE)


unmatchedPhiNames <- phiNames[is.na(namesToWBID)]
geneName <- geneNameAndWormBaseID$geneName
WBID <- geneNameAndWormBaseID$WBID




tidyJoinedData <- bind_rows(lifeStageCount, tidyFilteredAndWeightedPhiData) %>%
    mutate_if(is.character, as.factor)

joinedData <- pivot_wider(tidyJoinedData, names_from=stage, values_from=c(count, weight)) 

```
