---
title: "Fit ROC to Sequence Data in a FASTA File"
author: "Michael A. Gilchrist"
date: "06 Aug 2020"
output: pdf_document
---

# Background
- Use restart file from July 30 fit to start this one and run it for a long time with lots of thinning to get adquate sampling.
- Output from 30 July indicates ACF for CSP are still high at 40 samples (we used thinning = 1).
Recommend using thin=20.
- This code *doesn't work*.  I get the following errors
```
> sys.runtime<-system.time(
+ runMCMC(mcmc = mcmc, genome = genome, model = model, ncores = ncores, divergence.iteration = divergence.iteration)
+ )
Allowing divergence from initial conditions for 0 iterations.

maxGrouping: 22
Type: 
Starting MCMC
	Estimate Codon Specific Parameters? TRUE 
	Estimate Hyper Parameters? TRUE 
	Estimate Synthesis rates? TRUE 
	Starting MCMC with 40000 iterations
	Adapting will stop after 20000 steps

 *** caught segfault ***
address 0x56178db95d38, cause 'memory not mapped'

Traceback:
 1: .External(list(name = "CppMethod__invoke_void", address = <pointer: 0x5615440bdee0>,     dll = list(name = "Rcpp", path = "/usr/lib/R/site-library/Rcpp/libs/Rcpp.so",         dynamicLookup = TRUE, handle = <pointer: 0x5615450b84c0>,         info = <pointer: 0x561542eb0310>), numParameters = -1L),     <pointer: 0x561546de8810>, <pointer: 0x561545841060>, .pointer,     ...)
 2: mcmc$run(genome, model, ncores, divergence.iteration)
 3: runMCMC(mcmc = mcmc, genome = genome, model = model, ncores = ncores,     divergence.iteration = divergence.iteration)
 4: system.time(runMCMC(mcmc = mcmc, genome = genome, model = model,     ncores = ncores, divergence.iteration = divergence.iteration))

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection:
 *** caught segfault ***
address 0x561b54dd0438, cause 'memory not mapped'

Traceback:
 1: .External(list(name = "CppMethod__invoke_void", address = <pointer: 0x5615440bdee0>,     dll = list(name = "Rcpp", path = "/usr/lib/R/site-library/Rcpp/libs/Rcpp.so",         dynamicLookup = TRUE, handle = <pointer: 0x5615450b84c0>,         info = <pointer: 0x561542eb0310>), numParameters = -1L),     <pointer: 0x561546de8810>, <pointer: 0x561545841060>, .pointer,     ...)
[cheoah ~/Repositories/(4)/06_Fit.ROC.to.Sequences-Long.Fit.Using.Previous.Starting.Val
```

# ESS Commands
See (~/Software/R/r.notes.Rmd)  for more details

# Purpose
- Fit ROC to data using easier to understand and modify code
- Save output for future plotting.

# Load Libraries
```{r}

## Clear out any previously defined objects and parameters
rm(list=ls())

## Load Needed Libraries
library(tidyr)
library(tibble)
library(readr)
library(dplyr)
library(stringr)
library(forcats)
library(knitr)
library(purrr)
library(AnaCoDa, verbose = TRUE, lib.loc = "~/R/x86_64-pc-linux-gnu-library/3.6") ## Could use Version(); if it's not there build locally using script
## library(profmem, verbose = TRUE)  #is this still necessary?


saveImage <- TRUE ## Save image of session at end of each round of fitting?
```

# Local Items
## Functions
```{r}

## Use this function when creating (or recreating for plotting) genomes
myCreateGenomeObject <- function(fasta.file, gene.index = NULL){
    ## Make sure gene.index is a list, if it's not make it one.
    if( !is.list(gene.index) ) gene.index <- list(gene.index) 
    if( length(fasta.file) != length(gene.index) )
        stop('length(fasta.file) != length(gene.index)')

    n = length(fasta.file)
    
    previous.genome.size = 0  ## will be come a vector used to adjust index
    combined.index = c() ## will be come a vector of adjusted index
    ## Create unfiltered object combined across fasta.file
    for(i in 1:n){

        print(paste0("Using file ", fasta.file[i])) 
        if(i == 1){
            genome = NULL
            append = FALSE
        } else
        {
            genome = tmpGenome
            append = TRUE
        }
        
        tmpGenome <-
            initializeGenomeObject(
                file = fasta.file[i],
                genome = genome,
                match.expression.by.id = FALSE,
                append = append,
                positional = FALSE)  ## Positional is for panse model
        ## Create single sample.index for combined genome object
        updated.index <- gene.index[[i]] + previous.genome.size[i]
        combined.index <-c(combined.index, as.vector(updated.index))
        print(paste0("length(combined.index) = ", length(combined.index)) )

        previous.genome.size <- c(previous.genome.size, length(genome) )
        
    }

    
    ## Filter genome
    tmpGenome <- tmpGenome$getGenomeForGeneIndices(combined.index, simulated=FALSE)
    
    
    return(tmpGenome)
}

```

## File Settings
```{r}
input.dir <- "./Input/"  ## Location of FASTA and possibly empirical phi
output.dir <- "./Output/"
fasta.file <- paste0(input.dir, "c_elegans.PRJNA13758.WS277.CDS_transcripts.fa")

```

## Load and Redefine Some Parameters
```{r}


load("./Input/session.image.RData")

## Remove dangling objects
rm("mcmc", "genome", "parameter", "model")


   
previous.max.round  <- max.round


max.round <- 2 ## Max number of rounds of fitting
which.model="ROC"
which.genome = "monoform" # monoform or nuclear
sample.genome = FALSE #whether to create a subsample of the genome
set.seed = TRUE # flag for setting random seed
sample.seed = 2397502  # random seed for sampling genome
gene.sample.size <- 1500 ## number of genes to use in fitting
## Local settings
with.phi  <-  FALSE
init_sphi <- 1.8 ## initial value for s_phi
## Hyperparameters include s_phi, measurement error, and ??
est.hyper  <- TRUE ## FALSE keeps s_phi fixed
mixDef <- "allUnique"
runConvergenceTests <- FALSE

initial.divergence <- 0  ## # of divergence.iterations on round 1
samples <- c(2000,2000)  ## steps determined by samples * thin
thinning <- 20    ## only 1/thin steps (iterations) are kept (in memory?)
adaptive.width <- 100  ## adaptive width, in units of samples
steps <- samples * thinning 
## ncores <- (as.integer(system(command="nproc"))-1) #use n local cores -1
## could use `detectCores()` from library(parallel)
ncores <- 4 ## 3 or 4 is good for cheoah
adaptive.ratio <- c(0.5, 0) # proportion of samples to run with adaptive proposal widths
adaptive.steps <- steps * adaptive.ratio
samples.to.keep <- (1 - adaptive.ratio) * samples # Saves relative to last sample
    
## max.mixture <- 1
## mixture.sizes <- rep(0, max.mixture)
```

# ROC Functions (some)
```{r}
## Functions available for Objects
methods(class = "Rcpp_ROCParameter")
methods(class = "Rcpp_ROCModel")
methods(class = "Rcpp_ROCTrace") ## stored as part of parameter object
methods(class = "Rcpp_Genome")
methods(class = "Rcpp_MCMCAlgorithm")
```

# Set up GenomeObj
## Load Gene Lists
- Loads tibbles: nuclearGenes and monoisoformGenes
```{r}

##load("./Input/all.nuclear.and.monoform.gene.lists.Rdata")
names(nuclearGenes)
```

## Set Up sampleIndex (not needed)
Don't filter because sampleIndex is already defined.
```{r eval == False}
## create index for filtering genome
tmpGenes <- eval(str2lang(paste0(which.genome, "Genes")))
tmpIndex  <- tmpGenes$index ## Indexing is relative to order in full genome fasta file
## Should check to make sure ID's match indexes
## Or could revise getGenes to take IDs as well as indices
if(sample.genome == FALSE){
    sampleIndex  <- tmpIndex
} else
{
    if(set.seed == TRUE) set.seed(sample.seed)
    sampleIndex <- sort(sample(tmpIndex, gene.sample.size))
}

expectedIDs <- filter(tmpGenes, index %in% sampleIndex)$isoform.ID

length(expectedIDs)

```

### Create Genome Object
Create genome using sampleIndex
```{r}

genome <- myCreateGenomeObject(fasta.file= fasta.file, sampleIndex)


genomeIDs <- getNames(genome)  #really get IDs
genome.length <- length(genomeIDs)

matchID <- expectedIDs == genomeIDs

if(sum(matchID) == gene.sample.size) print("All isoform.IDs match!")
```

## Create Output Folders
```{r}


## Create output directories if needed
output.folder.list <- c("Graphs", "Restart", "Parameters", "R.Objects", "Run.Info")

tmp <- lapply(output.folder.list, function(x){
    tmpDir <- paste0(output.dir, x)
    if(!dir.exists(tmpDir)) dir.create(tmpDir, recursive = TRUE)
}
)

```

## Do Multiple Rounds of Model Fitting 
```{r}
round <- 1

#while(round <= max.round){

    input.restart.file <- paste0(output.dir,"Restart/rstart.round.", round - 1, ".rst")
    output.restart.file <- paste0(output.dir,"Restart/rstart.round.", round, ".rst")
    
    if(round==1){ #initial set up
        ## Initialize parameter object
        ## Set initial phi value
        input.restart.file <- paste0(input.dir,"Restart/rstart.round.", previous.max.round, ".rst")
    }
    
parameter <-
    initializeParameterObject(
        genome = genome,
        init.with.restart.file = input.restart.file,
        model=which.model)

divergence.iteration <- 0
    

    ## Initialize model object
model <- initializeModelObject(parameter = parameter, model = which.model, with.phi = with.phi)
    
    ## Save model object here?

    ## Initialize MCMC and Model Object
    mcmc <- initializeMCMCObject(samples=samples[round],
                                 thinning=thinning,
                                 adaptive.width=adaptive.width,
                                 est.expression=TRUE,
                                 est.csp=TRUE,
                                 est.hyper=est.hyper,
                                 est.mix = FALSE)

    ## Set steps for adaptive algorithm 
    mcmc$setStepsToAdapt(adaptive.steps[round])

    ## Set up restart files
setRestartSettings(mcmc = mcmc,
                   filename = output.restart.file,
                   samples = adaptive.width,
                   write.multiple = FALSE)  
    ## run mcmc on genome with parameter using model
sys.runtime<-system.time(
        runMCMC(mcmc = mcmc, genome = genome, model = model, ncores = ncores, divergence.iteration = divergence.iteration)
    )
    
    sys.runtime <- data.frame(Value=names(sys.runtime),Time=as.vector(sys.runtime))
    
    write.table(paste0(output.dir, "Run.Info/", sys.runtime),  #would like to add system.time to this.
                file=paste0(output.dir,"runtime.round-", round ,".csv"),
                sep=",",
                col.names = TRUE,
                row.names = TRUE,
                quote=FALSE)
    
    ## Process Output
    print("Begin Processing Output")
    print("Write Restart File")
    
    print("Write Summaries of CSP")
    ## Write output to .csv file
    getCSPEstimates(parameter = parameter,
                    filename = paste0(output.dir,
                                     "Parameters/csp.round-",
                                     round),
                    mixture = 1,
                    samples = samples.to.keep[round],
                    relative.to.optimal.codon= TRUE,
                    report.original.ref = TRUE)
    ##print("Save Genome Objects? No, they don't change with run.")
    print("Save Model Objects?")
    print("Save Parameter Object")
    output.file <- paste0(output.dir,"R.Objects/parameter.round-", round, ".Rda")
    writeParameterObject(parameter = parameter,
                         file = output.file     
                         )
    print("Save MCMC Object")
    writeMCMCObject(mcmc = mcmc,
                    file = paste0(output.dir,"R.Objects/mcmc.round-", round, ".Rda") )


    print("Save and Write Summaries of Phi to Memory")
    phi.vals <- getExpressionEstimates(
        parameter = parameter,
        gene.index = c(1:genome.length),
        samples = samples.to.keep[round])

    phi.tibble  <- bind_cols(isoform.ID = genomeIDs, phi.vals)
    write.table(phi.tibble,
                file=paste0(output.dir,"Parameters/phi.round-", round, ".csv"),
                sep=",",
                col.names = TRUE,
                quote = FALSE,
                row.names = FALSE)


    if(runConvergenceTests){
        ## Calculate & Export Diagnostics
        param.diag <- convergence.test(
            trace,
            samples=samples.to.keep[round],
            thin = thinning,
            what="Selection",
            mixture=1,
            frac1=0.1)
        z.scores <- param.diag$z[which(abs(param.diag$z) > 1.96)]
        if (length(z.scores) > 0)
        {
            param.conv <- FALSE
        }
        
        write(param.diag$z,
          paste0(output.dir,
                 "/Parameters/convergence.delta.eta.round-", round, ".txt"), ncolumns = 1)
    }else{
        print("not running convergence tests because 'no applicable method... applied to an object of class \"function\"")
    }

    if(saveImage) {
        ## save session to .RData
        ## Note: data in Rcpp objects 'mcmc', 'genome', 'parameter'
        ## are not saved properly!
        save.image(file = paste0(output.dir, "session.image.RData") )
        
        
    }
    round = round + 1
    

#}
```

